python-threatexchange/threatexchange/cli/cli_config.py

```diff
@@ -14,37 +8,251 @@
 a file (likely ~/.threatexchange)
 """
 
-FETCH_STATE_DIR_NAME = "fetched_state"
-COLLABORATION_CONFIG_DIR_NAME = "collaborations"
-INDEX_STATE_DIR_NAME = "index"
+from dataclasses import dataclass
+import sys
+import typing as t
+import json
+import pathlib
+import logging
+
+from dacite import WrongTypeError
+
+from threatexchange.fetcher import collab_config
+from threatexchange.fetcher.fetch_api import SignalExchangeAPI
+from threatexchange.content_type import content_base
+from threatexchange.fetcher.fetch_state import FetchedStateStoreBase
+from threatexchange.fetcher.apis.static_sample import StaticSampleSignalExchangeAPI
+from threatexchange.signal_type import signal_base
+from threatexchange.meta import FunctionalityMapping
+from threatexchange.cli.cli_state import CliSimpleState, CliIndexStore
+from threatexchange.cli import json as cli_json
+
+
 CONFIG_FILENAME = "config.json"
 
 
-class CliState:
+@dataclass
+class CLiConfig:
+    """A place to store misc configuration for the CLI"""
+
+    fb_threatexchange_api_token: t.Optional[str] = None
+
+
+class CliState(collab_config.CollaborationConfigStoreBase):
     """
     A wrapper around stateful information stored for the CLI.
 
     Everything is just in a single file (usually ~/.threatexchange).
     """
 
-    def __init__(self, dir: pathlib.Path):
-        assert dir.is_file
+    def __init__(
+        self, fetch_types: t.List[t.Union[SignalExchangeAPI, t.Type[SignalExchangeAPI]]]
+    ):
+        dir = pathlib.Path("~/.threatexchange/").expanduser()
         self._dir = dir
 
-    def get_collab_by_name(self, name: str) -> t.Optional[CollaborationConfig]:
+        self._name_to_ctype = {
+            ft.get_name(): ft.get_config_class() for ft in fetch_types
+        }
+
+        self._cache: t.Optional[
+            t.Dict[str, collab_config.CollaborationConfigBase]
+        ] = None
+
+        self._init_folders_if_needed()
+
+    def _init_folders_if_needed(self):
+        for d in (self.collab_dir, self.index_dir, self.fetched_state_dir):
+            if not d.is_dir():
+                d.mkdir(parents=True)
+        cfg = self.config_file
+        if not cfg.is_file():
+            cfg.write_text("{}")
+
+    @property
+    def collab_dir(self) -> pathlib.Path:
+        return self._dir / "collab_configs/"
+
+    @property
+    def fetched_state_dir(self) -> pathlib.Path:
+        return self._dir / "fetched"
+
+    @property
+    def index_dir(self) -> pathlib.Path:
+        return self._dir / "index/"
+
+    @property
+    def config_file(self) -> pathlib.Path:
+        return self._dir / "config.json"
+
+    def path_for_config(
```

Comment:
A lot of these issues are because this class inherits "CollabStoreBase" (which I tempted to delete), and also is pulling double duty as the general storage. 

I have a card for splitting up state access into a series of helpers where this will be easier to pry apart. I'd like to just TODO for now